src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/RandomExecutionGenerator.kt

```diff
@@ -0,0 +1,114 @@
+        val initExecution: MutableList<Actor> = ArrayList()
+        run {
+            var i = 0
+            while (i < testConfiguration.actorsBefore && validActorGeneratorsForInit.isNotEmpty()) {
+                val ag = validActorGeneratorsForInit[random.nextInt(validActorGeneratorsForInit.size)]
+                initExecution.add(ag.generate(0))
+                i++
+            }
+        }
+        // Create parallel execution part
+        // Construct non-parallel groups and parallel one
+        val nonParallelGroups = testStructure.operationGroups.stream()
+                .filter { g: OperationGroup -> g.nonParallel }
+                .collect(Collectors.toList())
+        nonParallelGroups.shuffle()
+        val parallelGroup: MutableList<ActorGenerator> = ArrayList(testStructure.actorGenerators)
+        nonParallelGroups.forEach(Consumer { g: OperationGroup -> parallelGroup.removeAll(g.actors) })
+        var parallelExecution: MutableList<MutableList<Actor>> = ArrayList()
+        val threadGens: MutableList<ThreadGen> = ArrayList()
+        for (t in 0 until testConfiguration.threads) {
+            parallelExecution.add(ArrayList())
+            threadGens.add(ThreadGen(t, testConfiguration.actorsPerThread))
+        }
+        for (i in nonParallelGroups.indices) {
+            threadGens[i % threadGens.size].nonParallelActorGenerators
+                    .addAll(nonParallelGroups[i]!!.actors)
+        }
+        val tgs2: List<ThreadGen> = ArrayList(threadGens)
+        while (threadGens.isNotEmpty()) {
+            val it = threadGens.iterator()
+            while (it.hasNext()) {
+                val threadGen = it.next()
+                val aGenIndexBound = threadGen.nonParallelActorGenerators.size + parallelGroup.size
+                if (aGenIndexBound == 0) {
+                    it.remove()
+                    continue
+                }
+                val aGenIndex = random.nextInt(aGenIndexBound)
+                val agen: ActorGenerator = if (aGenIndex < threadGen.nonParallelActorGenerators.size) {
+                    getActorGenFromGroup(threadGen.nonParallelActorGenerators, aGenIndex)
+                } else {
+                    getActorGenFromGroup(parallelGroup,
+                            aGenIndex - threadGen.nonParallelActorGenerators.size)
+                }
+                parallelExecution[threadGen.iThread].add(agen.generate(threadGen.iThread + 1))
+                if (--threadGen.left == 0) it.remove()
+            }
+        }
+        parallelExecution = parallelExecution.stream().filter { actors: List<Actor> -> actors.isNotEmpty() }.collect(Collectors.toList())
+        // Create post execution part if the parallel part does not have suspendable actors
+        val postExecution: MutableList<Actor>
+        if (parallelExecution.stream().noneMatch { actors: List<Actor> -> actors.stream().anyMatch(Actor::isSuspendable) }) {
```

Comment:
Please don't use Java's streams, we have a much simpler API in Kotlin without `stream()` invocation